src/_pytest/assertion/util.py

```diff
@@ -225,10 +227,21 @@ def _diff_text(left: str, right: str, verbose: int = 0) -> List[str]:
         left = repr(str(left))
         right = repr(str(right))
         explanation += ["Strings contain only whitespace, escaping them using repr()"]
-    explanation += [
-        line.strip("\n")
-        for line in ndiff(left.splitlines(keepends), right.splitlines(keepends))
-    ]
+
+    left_lines = left.splitlines(keepends)
+    right_lines = right.splitlines(keepends)
+
+    if any(
+        wcwidth(ch) <= -1
+        for ch in itertools.chain.from_iterable([x for x in left_lines + right_lines])
+    ):
+        left_lines = [repr(x) for x in left_lines]
+        right_lines = [repr(x) for x in right_lines]
+        explanation += [
+            "NOTE: Strings contain non-printable/zero-width characters. Escaping them using repr()."
```

Comment:
To keep the message short, I think the reference to zero-width characters can be removed. "Non-printable characters" does not seem to be a well-defined thing in Unicode, so we are free to include zero-width characters under that term. And it seems reasonable to do so to me...